freedomofpress/securedrop
# Local automated upgrade testing

## Status

Ready for review

## Description of Changes

Changes proposed in this pull request:

Fixes first check-box in #3018 

Add two scenarios:
* `vagrant_packager` -- builds two vagrant boxes ready for redistribution
* `upgrade` --- fires up boxes from the first scenario (pulled from s3), and chucks all locally built deps to a local apt server.


## Testing

How should the reviewer test this PR?

* `Caveat` - this PR testing will only work under a system with libvirt/kvm. There is another ticket to also add support for virtualbox. 

Packager testing:

* Run `make vagrant_package` - be prepared to enter a sudo password about 20 min in . 
* you'll find the vagrant box files in molecule/vagrant_packager/build/*.box - you can manually import those if you want to mess around with em
* Peek at `molecule/vagrant_packager/push.yml` dont run that now... but glance it over :)

Upgrade testing (local deb packages):
* `make build-debs`  -  make you sure have a version of debian packages that are higher than `0.6` if the `build/` directory. If you don't, running this command should give you `0.7.0~rc1` 
* Run `molecule converge -s upgrade` to get `.0.6` SD servers up. You'll get passed an onion address at the end. Navigate to that in a web-browser and note the version at the bottom.
* Run `molecule side-effect -s upgrade`  - Navigate back to the onion address and confirm version has bumped.

Upgrade testing (apt-test proxy):
1. Run `molecule converge -s upgrade` to get `.0.6` SD servers up.
2. Run `molecule login -s upgrade --host app-staging`  - To get a shell onto the app server
3. In the app server, run `sudo apt-get update` (confirm no ERRORS , you'll see a warning about duplicate repos but thats not an error and is a diff issue)
4. Run `apt-cache policy securedrop-config` (same terminal ^^) and you should see all the packages pulled in from above
5. Now back at the securedrop root dir run `QA_APTTEST=yes molecule converge -s upgrade -- --diff  -t apt`
6. Follow step `2 -> 4` again. This time you should see versions of securedrop-config pulled in from apt-test (this should have a few versions back and look like below):

```
vagrant@app-staging:~$ apt-cache policy securedrop-config                                                
securedrop-config:
  Installed: 0.1.0+0.6
  Candidate: 0.1.1+0.7.0~rc3
  Version table:
     0.1.1+0.7.0~rc3 0
        500 https://apt.freedom.press/ trusty/main amd64 Packages
     0.1.1+0.7.0~rc2 0
        500 https://apt.freedom.press/ trusty/main amd64 Packages
     0.1.1+0.7.0~rc1 0
        500 https://apt.freedom.press/ trusty/main amd64 Packages
 *** 0.1.0+0.6 0
        500 https://apt.freedom.press/ trusty/main amd64 Packages
        100 /var/lib/dpkg/status
     0.1.0+0.5.2 0
        500 https://apt.freedom.press/ trusty/main amd64 Packages
     0.1.0+0.5.1 0
        500 https://apt.freedom.press/ trusty/main amd64 Packages
```



## Deployment

Any special considerations for deployment? Consider both:

None - This only affects developer upgrade testing environment.

## Checklist

### If you made changes to the app code:

None

### If you made changes to the system configuration:

None

### If you made changes to documentation:

None
